supply/spec/uploader_spec.rb

```diff
@@ -203,25 +203,61 @@ def find_obbs
       end
     end
 
-    describe '#perform_upload' do
+    # add basic == functionality to LocalizedText class for testing purpose
+    class AndroidPublisher::LocalizedText
+      def ==(other)
+        self.language == other.language && self.text == other.text
+      end
+    end
+
+    shared_examples 'run supply to upload metadata' do |version_codes:, with_explicit_changelogs:|
+      let(:languages) { ['en-US', 'fr-FR', 'ja-JP'] }
+      subject(:release) { double('release', version_codes: version_codes) }
       let(:client) { double('client') }
-      let(:config) { { apk: 'some/path/app.apk' } }
+      let(:config) { { apk_paths: version_codes.map { |v_code| "some/path/app-v#{v_code}.apk" }, metadata_path: 'supply/spec/fixtures/metadata/android', track: 'track-name' } }
 
       before do
         Supply.config = config
         allow(Supply::Client).to receive(:make_from_config).and_return(client)
-        allow(client).to receive(:upload_apk).with(config[:apk]).and_return(1) # newly uploaded version code
+        version_codes.each do |version_code|
+          allow(client).to receive(:upload_apk).with("some/path/app-v#{version_code}.apk").and_return(version_code) # newly uploaded version code
+        end
+        allow(client).to receive(:upload_changelogs).and_return(nil)
+        allow(client).to receive(:tracks).with('track-name').and_return([double('tracks', releases: [ release ])])
+        languages.each do |lang|
+          allow(client).to receive(:listing_for_language).with(lang).and_return(Supply::Listing.new(client, lang))
+        end
         allow(client).to receive(:begin_edit).and_return(nil)
         allow(client).to receive(:commit_current_edit!).and_return(nil)
       end
 
-      it 'should update track with correct version codes' do
+      it 'should update track with correct version codes and optional changelog' do
         uploader = Supply::Uploader.new
-        expect(uploader).to receive(:update_track).with([1]).once
+        expect(uploader).to receive(:update_track).with(version_codes).once
+        version_codes.each do |version_code|
+          expected_notes = languages.map do |lang|
+            AndroidPublisher::LocalizedText.new(
+              language: lang,
+              text: "#{lang} changelog #{with_explicit_changelogs ? version_code : -1}"
+            )
+          end
```

Comment:
A side note that if one were to call the shared examples with `version_codes: [1,2], with_explicit_changelogs: false` then that part of the code would make `expected_notes` contain two entries with `#{lang} changelog -1` instead of just one.

In practice none of your tests calls it that way so that passes, but still an inconsistency in the test code to keep in mind.

One (verbose) way of fixing that logic would be this:
```suggestion
          expected_notes = if with_explicit_changelogs
            languages.map do |lang|
              AndroidPublisher::LocalizedText.new(
                language: lang,
                text: "#{lang} changelog version_code"
              )
            end
          else
            [
              AndroidPublisher::LocalizedText.new(
                language: lang,
                text: "#{lang} changelog -1"
              )
            ]
          end
```

Another way would be this (thanks to `LocalizedText#==` you added above):
```suggestion
          expected_notes = languages.map do |lang|
            AndroidPublisher::LocalizedText.new(
              language: lang,
              text: "#{lang} changelog #{with_explicit_changelogs ? version_code : -1}"
            )
          end.uniq
```

Reply:
Thanks for the suggestion! (and the reviews so far 😄). Tested that and added the `uniq` 👌🏻 

As soon as the build is 🟢 I'll merge this PR.